src/manifest/pfm.rs

```diff
@@ -419,6 +440,549 @@ impl<'pfm, F: Flash> Pfm<'pfm, F> {
+    /// and potentially allocating it on `arena`.
+    pub fn read<'buf>(
+        self,
+        sha: &impl sha256::Builder,
+        arena: &'buf impl Arena,
+    ) -> Result<FwVersion<'a, 'buf, 'pfm, F>, Error>
+    where
+        'pfm: 'buf,
+    {
+        // We can't avoid this read due to the requirement of both verifying
+        // the hash and having all of the data read into memory at once.
+        //
+        // Note that this needs to be aligned to a 4-byte boundary for some of
+        // the zero-copy operations below to work.
+        let mut buf = &*self.pfm.flash.read_direct(
+            Region::new(
+                self.toc_entry.offset() as u32,
+                self.toc_entry.len() as u32,
+            ),
+            arena,
+            mem::align_of::<u32>(),
+        )?;
+        if let Some(expected) = self.hash {
+            let mut hash = [0; 32];
+            sha.hash_contiguous(buf, &mut hash)?;
+            if &hash != expected {
+                return Err(Error::SignatureFailure);
+            }
+        }
+
+        let image_count = buf.read_le::<u8>()? as usize;
+        let rw_count = buf.read_le::<u8>()? as usize;
+        let version_len = buf.read_le::<u8>()? as usize;
+        let _ = buf.read_le::<u8>()?;
+        let version_addr = buf.read_le::<u32>()?;
+
+        if buf.len() < version_len {
+            return Err(Error::OutOfRange);
+        }
+        let (version_str, mut buf) = buf.split_at(version_len);
+
+        // Align back to 4-byte boundary.
+        buf = &buf[misalign_of(buf.as_ptr() as usize, 4)..];
+
+        let rw_len = mem::size_of::<RwRegion>() * rw_count;
+        let (rw_bytes, unparsed_image_regions) = buf.split_at(rw_len);
+        // NOTE: This cannot panic, since it checks for alignment (4-byte) and
+        // size, which have already been explicitly checked above.
+        let rw_regions = LayoutVerified::<_, [RwRegion]>::new_slice(rw_bytes)
+            .unwrap()
+            .into_slice();
+        for rw in rw_regions {
+            if rw.start_addr > rw.end_addr {
+                return Err(Error::OutOfRange);
+            }
+        }
+
+        let image_region_offsets = arena.alloc_slice::<u32>(image_count)?;
+        if image_count > 0 {
+            // FIXME: we don't need to actually track this "first" offset.
+            image_region_offsets[0] = 0;
+            for i in 0..image_count {
+                let rest =
+                    &unparsed_image_regions[image_region_offsets[i] as usize..];
+                let (header, rest) =
+                    match LayoutVerified::<_, FwRegionHeader>::new_from_prefix(
+                        rest,
+                    ) {
+                        Some(h) => h,
+                        None => return Err(Error::OutOfRange),
+                    };
+                // FIXME: we don't deal with hash types that aren't SHA-256.
+                if header.hash_type != 0 {
```

Comment:
What about an enum to represent the hash type? Just `!= 0` seems inline-magic-valuey.